### PR TITLE
Add support for using mingw64 clang on MS-Windows to build Vim

### PR DESCRIPTION
**Pull request by yegappan**

### CODE REVIEW

**Comment by brammool** [comment]
any comments?

**Comment by k-takata** [comment]
Recent MSYS2 provides two types of Clang compilers:

1. Installed in the `/mingw{64,32}/bin` directory. Usable via the `C:\msys64\mingw{64,32}.exe` shell. (In my understanding,) ASAN is **not** available.
   Packages: `mingw-w64-{x86_64,i686}-clang`, etc.
2. Installed in the `/clang{64,32}/bin` directory. Usable via the `C:\msys64\clang{64,32}.exe` shell. ASAN is available.
   Packages: `mingw-w64-clang-{x86_64,i686}-clang`, etc.

See https://www.msys2.org/docs/environments/ for detail.

If one wants to use clang on MSYS2, the CLANG{64,32} environments might be better.

File: src/INSTALLpc.txt [hunk 1]
```diff
@@ -282,6 +282,27 @@ This command is in msys32.bat.  Or for the 64 bit compiler use msys64.bat:
...
+
+* mingw-w64-x86_64-clang
+
+Use the following command to install it:
+
+    $ pacman -S mingw-w64-x86_64-clang
+
+Go to the source directory of Vim, then execute the make command.  E.g.:
+
+    CC=clang
```
Comment (k-takata) [inline src/INSTALLpc.txt:297]: Isn't `CXX=clang++` needed when `OLE` or `DIRECTX` is set to yes?

**Comment by vim-ml** [comment]
Hi,

On Sun, Jun 12, 2022 at 7:52 PM K.Takata ***@***.***> wrote:

> Recent MSYS2 provides two types of Clang compilers:
>
>    1. Installed in the /mingw{64,32}/bin directory. Usable via the
>    C:\msys64\mingw{64,32}.exe shell. (In my understanding,) ASAN is *not*
>    available.
>    Packages: mingw-w64-{x86_64,i686}-clang, etc.
>
>
I have used the clang installed in the \mingw64\bin directory to build Vim
(from the command-line shell and not
from the msys2 environment).  I am able to build Vim with ASAN with this
compiler.


>
>    1. Installed in the /clang{64,32}/bin directory. Usable via the
>    C:\msys64\clang{64,32}.exe shell. ASAN is available.
>    Packages: mingw-w64-clang-{x86_64,i686}-clang, etc.
>
> See https://www.msys2.org/docs/environments/ for detail.
>
> If one wants to use clang on MSYS2, the CLANG{64,32} environments might be
> better.
>
>
>
I am also able to build Vim with the MSYS2 CLANG64 environment.

Regards,
Yegappan

**Comment by vim-ml** [comment]
Hi,

On Sun, Jun 12, 2022 at 7:56 PM K.Takata ***@***.***> wrote:
>
> @k-takata commented on this pull request.
>
> ________________________________
>
> In src/INSTALLpc.txt:
>
> > @@ -282,6 +282,27 @@ This command is in msys32.bat.  Or for the 64 bit compiler use msys64.bat:
>  If you have msys64 in another location you will need to adjust the paths for
>  that.
>
> +2.5. Build Vim with Clang
> +
> +The following package group is required for building Vim with Clang:
> +
> +* mingw-w64-x86_64-clang
> +
> +Use the following command to install it:
> +
> +    $ pacman -S mingw-w64-x86_64-clang
> +
> +Go to the source directory of Vim, then execute the make command.  E.g.:
> +
> +    CC=clang
>
> Isn't CXX=clang++ needed when OLE or DIRECTX is set to yes?
>

Yes. I have updated the PR.

Thanks,
Yegappan

**Comment by k-takata** [comment]
> I am able to build Vim with ASAN with this
compiler.

That is nice. When I asked the msys2 team last year, ASAN was available only on the clang{64,32} environment:
https://github.com/msys2/MINGW-packages/issues/8476

**Comment by k-takata** [comment]
Hmm, I couldn't build with the MINGW64 version of clang:

```
$ clang++ -O1 -g -fsanitize=address -fno-omit-frame-pointer example_UseAfterFree.cc
C:/msys64/mingw64/bin/ld: cannot find C:/msys64/mingw64/lib/clang/14.0.4/lib/windows/libclang_rt.asan_dynamic-x86_64.dll.a: No such file or directory
C:/msys64/mingw64/bin/ld: cannot find C:/msys64/mingw64/lib/clang/14.0.4/lib/windows/libclang_rt.asan_dynamic_runtime_thunk-x86_64.a: No such file or directory
C:/msys64/mingw64/bin/ld: cannot find C:/msys64/mingw64/lib/clang/14.0.4/lib/windows/libclang_rt.asan_dynamic_runtime_thunk-x86_64.a: No such file or directory
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```
`example_UseAfterFree.cc` is coming from the official document of clang:
https://clang.llvm.org/docs/AddressSanitizer.html

**Pull request closed without merging**